psydac/linalg/kron.py

```diff
@@ -325,170 +197,490 @@ def transpose(self):
+        and b is a suitable vector.
+        """
 
-    for i2 in range(e2-s2+1):
-        for i1 in range(e1-s1+1):
-            Y_loc = Y[i1, i2, :]  # 1D contiguous slice
-            subcomm_3.Allgatherv( Y_loc, [Y_glob_3, sizes3, disps3, mpi_type] )
-            Y[i1, i2, :] = A3.solve( Y_glob_3 )[s3:e3+1]
+        # type checks
+        assert rhs.space is self._space
 
-    # ...
-    out.update_ghost_regions()
-    # ...
+        if out is not None:
+            assert isinstance( out, StencilVector )
+            assert out.space is self._space
+        else:
+            out = StencilVector( rhs.space )
+        
+        inslice = rhs[self._slice]
+        outslice = out[self._slice]
 
-    return out
+        # call the actual kernel
+        self._solve_nd(inslice, outslice, transposed)
+        
+        out.update_ghost_regions()
+        return out
+ 
+    def _solve_nd(self, inslice, outslice, transposed):
+        """
+        The internal solve loop. Can handle arbitrary dimensions.
+        """
+        temp1 = self._temp1
+        temp2 = self._temp2
+
+        # copy input
+        self._inslice_to_temp(inslice, temp1)
+
+        # internal passes
+        for i in range(self._ndim - 1):
+            # solve direction
+            self._solver_passes[i].solve_pass(temp1, temp2, transposed)
+
+            # reorder and swap
+            self._reorder_temp_to_temp(temp1, temp2, i)
+            temp1, temp2 = temp2, temp1
+        
+        # last pass
+        self._solver_passes[-1].solve_pass(temp1, temp2, transposed)
+
+        # copy to output
+        self._reorder_temp_to_outslice(temp1, outslice)
+
+    def _inslice_to_temp(self, inslice, target):
+        """
+        Copies data to an internal, 1-dimensional temporary array.
+        Does not allocate any new array.
+        """
+        targetview = target[:self._localsize]
+        targetview.shape = inslice.shape
+
+        targetview[:] = inslice
+    
+    def _reorder_temp_to_temp(self, source, target, i):
+        """
+        Reorders the dimensions of the temporary arrays, and copies data from one to another.
+        Does not allocate any new array.
+        """
+        sourceview = source[:self._localsize]
+        sourceview.shape = self._shapes[i]
+
+        targetview = target[:self._localsize]
+        targetview.shape = self._shapes[i+1]
+
+        targetview[:] = sourceview.transpose(self._perm)
+    
+    def _reorder_temp_to_outslice(self, source, outslice):
+        """
+        Reorders the dimensions of the temporary array for a final time, and copies it to the output.
+        Does not allocate any new array.
+        """
+        sourceview = source[:self._localsize]
+        sourceview.shape = self._shapes[-1]
+
+        outslice[:] = sourceview.transpose(self._perm)
+
+    class KroneckerSolverSerialPass:
+        """
+        Solves several linear equations at the same time,
+        given that the data is already in memory.
```

Comment:
This docstring should be clarified: the same equation is solved for multiple right-hand-side vectors simultaneously.